GameServerLib/API/ApiEventManager.cs

```diff
@@ -66,158 +75,418 @@ internal static void SetGame(Game game)
+        public static EventOnSpellCast OnSpellCast = new EventOnSpellCast();
+        public static EventOnSpellChannel OnSpellChannel = new EventOnSpellChannel();
+        public static EventOnSpellChannelCancel OnSpellChannelCancel = new EventOnSpellChannelCancel();
+        public static EventOnSpellHit OnSpellHit = new EventOnSpellHit();
+        public static EventOnSpellPostCast OnSpellPostCast = new EventOnSpellPostCast();
+        public static EventOnSpellPostChannel OnSpellPostChannel = new EventOnSpellPostChannel();
+        public static EventOnTakeDamage OnTakeDamage = new EventOnTakeDamage();
+        // TODO: Handle crowd control the same as normal dashes.
         public static EventOnUnitCrowdControlled OnUnitCrowdControlled = new EventOnUnitCrowdControlled();
-        public static EventOnChampionCrowdControlled OnChampionCrowdControlled = new EventOnChampionCrowdControlled();
+        // TODO: Change to OnMoveSuccess and change where Publish is called internally to reflect the name.
+        public static EventOnUnitUpdateMoveOrder OnUnitUpdateMoveOrder = new EventOnUnitUpdateMoveOrder();
     }
 
-    public class EventOnUpdate
+    public class EventOnHitUnit
     {
-        private List<Tuple<object, Action<float>>> _listeners = new List<Tuple<object, Action<float>>>();
-        public void AddListener(object owner, Action<float> callback)
+        private readonly List<Tuple<object, IObjAiBase, Action<IAttackableUnit, bool>, bool>> _listeners = new List<Tuple<object, IObjAiBase, Action<IAttackableUnit, bool>, bool>>();
+        public void AddListener(object owner, IObjAiBase unit, Action<IAttackableUnit, bool> callback, bool singleInstance)
         {
-            var listenerTuple = new Tuple<object, Action<float>>(owner, callback);
+            var listenerTuple = new Tuple<object, IObjAiBase, Action<IAttackableUnit, bool>, bool>(owner, unit, callback, singleInstance);
             _listeners.Add(listenerTuple);
         }
-
+        public void RemoveListener(object owner, IObjAiBase unit)
+        {
+            _listeners.RemoveAll((listener) => listener.Item1 == owner && listener.Item2 == unit);
+        }
         public void RemoveListener(object owner)
         {
-            _listeners.RemoveAll(listener => listener.Item1 == owner);
+            _listeners.RemoveAll((listener) => listener.Item1 == owner);
         }
-
-        public void Publish(float diff)
+        public void Publish(IObjAiBase unit, IAttackableUnit target, bool isCrit)
         {
-            _listeners.ForEach(listener => listener.Item2(diff));
+            var count = _listeners.Count;
+
+            if (count == 0)
+            {
+                return;
+            }
+
+            // TODO: Replace this method with a single function and just pass in count as a parameter (do this for all events which use this method).
+            for (int i = count; i > 0; i--)
```

Comment:
Here it should be `count-1`. Also, the condition should be `>=0` to avoid missing the first index.